pkg/storage/utils/decomposedfs/upload.go

```diff
@@ -684,83 +701,189 @@ func (upload *fileUpload) FinishUpload(ctx context.Context) (err error) {
-			/*
-				return strings.HasPrefix(attributeName, xattrs.GrantPrefix) || // for grants
-					strings.HasPrefix(attributeName, xattrs.MetadataPrefix) || // for arbitrary metadata
-					strings.HasPrefix(attributeName, xattrs.FavPrefix) || // for favorites
-					strings.HasPrefix(attributeName, xattrs.SpaceNameAttr) || // for a shared file
-			*/
-		})
+		// touch version node
+		file, err := os.Create(versionsPath)
 		if err != nil {
-			sublog.Info().Err(err).Msg("Decomposedfs: failed to copy xattrs")
+			discardBlob()
+			sublog.Err(err).Str("version", versionsPath).Msg("could not create version node")
+			return errtypes.InternalError("could not create version node")
+		}
+		defer file.Close()
+
+		fi, err := file.Stat()
+		if err != nil {
+			discardBlob()
+			sublog.Err(err).Str("version", versionsPath).Msg("could not stat version node")
+			return errtypes.InternalError("could not stat version node")
+		}
+		newMtime = fi.ModTime()
+
+		// copy blob metadata to version node
+		err = xattrs.CopyMetadataWithSourceLock(targetPath, versionsPath, func(attributeName string) bool {
+			return strings.HasPrefix(attributeName, xattrs.ChecksumPrefix) ||
+				attributeName == xattrs.BlobIDAttr ||
+				attributeName == xattrs.BlobsizeAttr
+		}, lock)
+		if err != nil {
+			discardBlob()
+			sublog.Err(err).Str("version", versionsPath).Msg("failed to copy xattrs to version node")
+			return errtypes.InternalError("failed to copy blob xattrs to version node")
+		}
+
+		// keep mtime from previous version
+		if err := os.Chtimes(versionsPath, oldMtime, oldMtime); err != nil {
+			discardBlob()
+			sublog.Err(err).Str("version", versionsPath).Msg("failed to change mtime of version node")
+			return errtypes.InternalError("failed to change mtime of version node")
 		}
-	}
 
-	// now try write all checksums
-	tryWritingChecksum(&sublog, n, "sha1", sha1h)
-	tryWritingChecksum(&sublog, n, "md5", md5h)
-	tryWritingChecksum(&sublog, n, "adler32", adler32h)
+		// we MUST bypass any cache here as we have to calculate the size diff atomically
+		oldSize, err = node.ReadBlobSizeAttr(targetPath)
+		if err != nil {
+			discardBlob()
+			sublog.Err(err).Str("version", versionsPath).Msg("failed to read old blobsize")
+			return errtypes.InternalError("failed to read old blobsize")
+		}
+	} else {
+		// touch metadata node
+		file, err := os.Create(targetPath)
+		if err != nil {
+			discardBlob()
+			sublog.Err(err).Msg("could not create node")
+			return errtypes.InternalError("could not create node")
+		}
+		file.Close()
 
-	// who will become the owner?  the owner of the parent actually ... not the currently logged in user
-	err = n.WriteAllNodeMetadata()
+		// basic node metadata
+		attrs[xattrs.ParentidAttr] = n.ParentID
+		attrs[xattrs.NameAttr] = n.Name
+		oldSize = 0
+	}
+
+	// update node metadata with new blobid etc
+	err = n.SetXattrsWithLock(attrs, lock)
```

Comment:
this `n` was defined fife or six screens up (for me, not on 4k) - I have an idea that it might be a node. Maybe the node of the new file? Maybe the node of the parent dir? (why not?) - or maybe not a Node, but a file counter called n? 
I remember times where C programmers thought that it is efficient to use one char vars which turned out to be a stupid idea that was expensive and annoying... I am really puzzled that we repeat that today. :sheep: